cpp/dolfin/mesh/Mesh.cpp

```diff
@@ -227,25 +228,18 @@ Mesh::Mesh(
         "Cannot create mesh. Wrong number of global cell indices");
   }
 
-  // Permutation from VTK to DOLFIN order for cell geometric nodes
-  std::vector<std::uint8_t> cell_permutation;
-  if (type == mesh::CellType::quadrilateral)
-  {
-    // Quadrilateral cells does not follow counter clockwise
-    // order (cc), but lexiographic order (LG). This breaks the assumptions
-    // that the cell permutation is the same as the VTK-map.
-    if (num_vertices_per_cell == cells.cols())
-      cell_permutation = {0, 1, 2, 3};
-    else
-      throw std::runtime_error("Higher order quadrilateral not supported");
-  }
-  else
-    cell_permutation = mesh::vtk_mapping(type, cells.cols());
-
   // Find degree of mesh
   // FIXME: degree should probably be in MeshGeometry
   _degree = mesh::cell_degree(type, cells.cols());
 
+  // Permutation from Mesh input format to DOLFIN order for cell geometric nodes
+  std::vector<std::uint8_t> cell_permutation;
+  if (custom_permutation.size() == 0)
+    cell_permutation = mesh::default_cell_permutation(type, _degree);
```

Comment:
Is this needed?

Reply:
We will assume that the user does not supply the permutation, and then we need to deduce it from the element type and degree.

Reply:
If not permutation is required it would be better that we apply no permutation. We could be explicit and pass in the cell type (or maybe just tdim and gdim?).

The less manipulation of data behind-the-scenes the better.

Reply:
No permutation means that the input mesh has to be ordered as UFC would like it. This is rarely the case.This `if-test` only checks if the user has supplied a permutation, if not, it assumes that it follows our default permutations, which right now is VTK for triangles, tetrahedrons, higher order quads, and lexiographic for first order quadrilaterals and hexahedrons.

Reply:
1. I've now added documentation of the ordering (In as brief manner as possible.
2. There is a free function for this, `vtk_mapping`.

Reply:
@jorgensd @garth-wells I think we agreed to rename UFC interface as FEniCS interface (working on a PR), so I'd rather say it is a FEniCS ordering.